### ISSUE TITLE
[Backend] Direct Calls Optimization

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#538]
As a special case, when we have an `ApplyDef(f, args)` node, we usually generate:

```scala
Fn1$Int$Int f = g.Root.f
f.setArg0(42);
g.next = f;

Fn1$Int$Int fn = null;
do {
  fn = g.next;
  g.next = null;
  fn.apply();
} while (g.next != null)
fn.getResult();
```

However, as an optimization, if we are willing to generate 2x the code, we could have a version of `f` that takes its arguments as an ordinary static Java method. E.g. we would have:

```java
class Root {
  public static int f(Global g, int x) {}
}
```

and we would generate an ordinary call:

```java
Root.f(g, 42) // static call
```

Note however that we must still pass `g` into `f` s.t. `f` can perform its own tail calls.

**Comment by magnus-madsen** [comment]
Let us assume we will do this by default and I will add a flag `-Xno-direct-calls` to disable this optimization.

**Comment by magnus-madsen** [comment]
Fixed #594
